Add metadata types and generator lib to repo

GavinRay97 opened the pull request:
<!-- Thank you for submitting this PR! :) -->
<!-- Provide a general summary of your changes in the Title above ^, end with (close #<issue-no>) or (fix #<issue-no>) -->

### Description
<!-- The title might not be enough to convey how this change affects the user. -->
<!-- Describe the changes from a user's perspective -->

Merges the metadata types and generator lib into main repo.
Console and CLI can now import and use these metadata types. Server can too, though those types probably already exist.

See `README.md` and `RFC.md` for more info

### Changelog

- [ ] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components
<!-- Remove non-affected components from the list -->

- [ ] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System
- [ ] Tests
- [x] Other (list it)

### Related Issues
<!-- Please make sure you have an issue associated with this Pull Request -->
<!-- And then add `(close #<issue-no>)` to the pull request title -->
<!-- Add the issue number below (e.g. #234) -->

### Solution and Design
<!-- How is this issue solved/fixed? What is the design? -->
<!-- It's better if we elaborate -->

### Steps to test and verify
<!-- If this is a feature, what are the steps to try them out? -->
<!-- If this is a bug-fix, how do we verify the fix? -->

### Limitations, known bugs & workarounds
<!-- Limitations of the PR, known bugs and suggested workarounds -->
<!-- Feel free to delete these comment lines -->

### Server checklist
<!-- A checklist for server code -->

#### Catalog upgrade
<!-- Is hdb_catalog version bumped? -->
Does this PR change Hasura Catalog version?
- [x] No
- [ ] Yes
  - [ ] Updated docs with SQL for downgrading the catalog <!-- https://hasura.io/docs/1.0/graphql/manual/deployment/downgrading.html#downgrading-across-catalogue-versions -->

#### Metadata
<!-- Hasura metadata changes -->

Does this PR add a new Metadata feature?
- [x] No
- [ ] Yes
  - Does `run_sql` auto manages the new metadata through schema diffing?
    - [ ] Yes
    - [ ] Not required
  - Does `run_sql` auto manages the definitions of metadata on renaming?
    - [ ] Yes
    - [ ] Not required
  - Does `export_metadata`/`replace_metadata` supports the new metadata added?
    - [ ] Yes
    - [ ] Not required


#### GraphQL
- [x] No new GraphQL schema is generated
- [ ] New GraphQL schema is being generated:
   - [ ] New types and typenames are correlated
   <!-- No dangling types or typenames with missing types (a potential bug, introspection fails) -->
   <!-- If you have anything in your mind, which can be added here as a check list item, please submit a PR to update this template :) -->

#### Breaking changes

- [x] No Breaking changes
- [ ] There are breaking changes:

  1. Metadata API

     Existing `query` types:
     - [ ] Modify `args` payload which is not backward compatible
     - [ ] Behavioural change of the API
     - [ ] Change in response `JSON` schema
     - [ ] Change in error code
     <!-- Add if anything not listed above -->

  2. GraphQL API

     Schema Generation:
     <!-- Any changes in schema auto-generation logic -->
     <!-- All GraphQL schema names are case sensitive -->
     - [ ] Change in any `NamedType`
     - [ ] Change in table field names
     <!-- Add if anything not listed above -->

     Schema Resolve:-
     <!-- Any change in logic of resolving input request -->
     - [ ] Change in treatment of `null` value for any input fields <!-- Explain them below -->
     <!-- Add if anything not listed above -->

  3. Logging

     - [ ] Log `JSON` schema has changed
     - [ ] Log `type` names have changed
     <!-- Add if anything not listed above -->

<!-- Add any other breaking change not mentioned above -->

<!-- Explain briefly about your breaking changes below -->

beerose commented on metadata-types/tsconfig.json:
```diff
@@ -0,0 +1,66 @@
+    // "isolatedModules": true,               /* Transpile each file as a separate module (similar to 'ts.transpileModule'). */
+
+    /* Strict Type-Checking Options */
+    "strict": false /* Enable all strict type-checking options. */,
```
Erm...

GavinRay97 replied:
Enabled strict

beerose commented on metadata-types/tsconfig.json:
```diff
@@ -0,0 +1,66 @@
+{
+  "compilerOptions": {
```
Can you remove commented lines?

GavinRay97 replied:
Done :+1:

nksaraf commented on metadata-types/src/types/HasuraMetadataV2.ts:
```diff
@@ -0,0 +1,828 @@
+  /** Comment */
+  comment?: string
+  /** Permissions of the action */
+  permissions?: { role: string }
```
Hi I was using the types in a project, I think this should be 
```tsx
    permissions?: { role: string }[]
```
This is to give multiple roles permissions for an action

GavinRay97 replied:
Ahh good catch, thank you. Will fix today. Need to get more sample `metadata.json` to put in the `tests` directory to catch this.

https://github.com/GavinRay97/graphql-engine/tree/feature/add-metadata-types-and-generator/metadata-types/src/tests

GavinRay97 commented:
> A couple of suggestions and questions.
> 
> 1. `Typescript` -> `TypeScript`.
> 2. Is there any reason why we don't have strict TypeScript?
> 3. Does adding minimal eslint and prettier config make sense?

Enabled strict TS, Find + Replace All for `Typescript` -> `TypeScript`, and added `.prettierrc`.
Fixed above comment about `Action.permission` needing array.

beerose commented:
Are we going to rename it to `metadata-sdk`?

beerose commented on metadata-types/package.json:
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@hasura/metadata",
```
```suggestion
  "name": "",
```

tirumaraiselvan commented:
@GavinRay97 Could you add a changelog pls?

tirumaraiselvan reviewed (approved):
changelog

tirumaraiselvan commented on CHANGELOG.md:
```diff
@@ -62,7 +100,7 @@ This release contains the [PDV refactor (#4111)](https://github.com/hasura/graph
 - cli: add missing global flags for seeds command (#5565)
 - docs: add docs page on networking with docker (close #4346) (#4811)
-
+- sdk: add `/contrib` top-level folder and Metadata Types SDK in contrib (#5452)
```
There is a conflict in changelog because of other PRs getting merged. When you fix it, could you also remove this line as it has been covered in highlights.
